service/history/api/updateworkflowoptions/api.go

```diff
@@ -0,0 +1,160 @@
+	ret := &historyservice.UpdateWorkflowExecutionOptionsResponse{}
+
+	err = api.GetAndUpdateWorkflowWithNew(
+		ctx,
+		nil,
+		definition.NewWorkflowKey(
+			ns.ID().String(),
+			req.GetWorkflowExecution().GetWorkflowId(),
+			req.GetWorkflowExecution().GetRunId(),
+		),
+		func(workflowLease api.WorkflowLease) (_ *api.UpdateWorkflowAction, updateError error) {
+			mutableState := workflowLease.GetMutableState()
+			defer func() { workflowLease.GetReleaseFn()(updateError) }()
+
+			if !mutableState.IsWorkflowExecutionRunning() {
+				// in-memory mutable state is still clean, let updateError=nil in the defer func()
+				// to prevent clearing and reloading mutable state while releasing the lock
+				return nil, consts.ErrWorkflowCompleted
+			}
+
+			// TODO (carly): dedupe by requestID
+			_ = req.GetRequestId()
+
+			// Merge the requested options mentioned in the field mask with the current options in the mutable state
+			mergedOpts, updateError := MergeOptions(
+				req.GetUpdateMask().GetPaths(),
+				req.GetWorkflowExecutionOptions(),
+				GetOptionsFromMutableState(mutableState),
+			)
+			if updateError != nil {
+				return nil, updateError
+			}
+
+			// Set options for gRPC response
+			ret.WorkflowExecutionOptions = mergedOpts
+
+			// If there is no mutable state change at all, return with no new history event and Noop=true
+			if proto.Equal(mergedOpts, GetOptionsFromMutableState(mutableState)) {
+				return &api.UpdateWorkflowAction{
+					Noop:               true,
+					CreateWorkflowTask: false,
+				}, nil
+			}
+
+			_, updateError = mutableState.AddWorkflowExecutionOptionsUpdatedEvent(req.GetWorkflowExecutionOptions().GetVersioningOverride())
+			if updateError != nil {
+				return nil, updateError
+			}
+
+			// TODO (carly) part 2: handle safe deployment change --> CreateWorkflowTask=true
+			return &api.UpdateWorkflowAction{
+				Noop:               false,
+				CreateWorkflowTask: false,
+			}, nil
+		},
+		nil,
+		shardCtx,
+		workflowConsistencyChecker,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+func GetOptionsFromMutableState(ms workflow.MutableState) *workflowpb.WorkflowExecutionOptions {
+	opts := &workflowpb.WorkflowExecutionOptions{}
+	// todo (carly) or todo (shahab): Have VersioningInfo store VersioningOverride instead of DeploymentOverride + BehaviorOverride separately
+	if versioningInfo := ms.GetExecutionInfo().GetVersioningInfo(); versioningInfo != nil {
+		if behaviorOverride := versioningInfo.GetBehaviorOverride(); behaviorOverride != enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED {
+			opts.VersioningOverride = &workflowpb.VersioningOverride{
+				Behavior:   behaviorOverride,
+				Deployment: versioningInfo.GetDeploymentOverride(),
+			}
+		}
+	}
+	return opts
+}
+
+// MergeOptions copies the given paths in `src` struct to `dst` struct
+func MergeOptions(paths []string, src, dst *workflowpb.WorkflowExecutionOptions) (*workflowpb.WorkflowExecutionOptions, error) {
```

Comment:
Did you mock it from activity options?

Reply:
no, because UpdateActivityOptions is not implemented in the versioning-3 branch, but I see Yuri wrote a util for this, so I'll adjust to use that

Reply:
I just reimplemented it based on the stuff Yuri wrote